### ISSUE TITLE
Rest of chain is lost after a task is retried chained to pre-work (another task)

### ISSUE THREAD

**Issue by nbrede** [celery/celery#6631]
# Checklist

- [x] I have verified that the issue exists against the `master` branch of Celery.
- [ ] This has already been asked to the [discussion group](https://groups.google.com/forum/#!forum/celery-users) first.
- [x] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [x] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.
- [x] I have included the contents of ``pip freeze`` in the issue.
- [x] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- None

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**: 5.0.5 (singularity)

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
software -> celery:5.0.5 (singularity) kombu:5.0.2 py:3.7.8
            billiard:3.6.3.0 filesystem:N/A
platform -> system:Windows arch:64bit, WindowsPE
            kernel version:10 imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:filesystem results:file:///\

broker_url: 'filesystem://localhost//'
result_backend: 'file:///\\'
result_expires: 900
accept_content: ['json']
task_serializer: 'json'
task_queues: [<some queues>]
task_create_missing_queues: False
task_default_queue: 'celery'
timezone: 'UTC'
enable_utc: True
worker_hijack_root_logger: False
worker_prefetch_multiplier: 1
broker_connection_timeout: 30
broker_transport_options: {
    'data_folder_in': '<temp path>',
    'data_folder_out': '<temp path>'}
task_send_sent_event: True
worker_send_task_events: True
deprecated_settings: None
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

* **Minimal Python Version**: 3.7.x 
* **Minimal Celery Version**: 5.0.5 (also happens with 4.4.x)
* **Minimal Kombu Version**: 4.6.11 (also happens with 5.0.2)
* **Minimal Broker Version**: N/A or Unknown
* **Minimal Result Backend Version**: N/A or Unknown
* **Minimal OS and/or Kernel Version**: N/A or Unknown
* **Minimal Broker Client Version**: N/A or Unknown
* **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
amqp==5.0.5
billiard==3.6.3.0
celery==5.0.5
click==7.1.2
click-didyoumean==0.0.3
click-plugins==1.1.1
click-repl==0.1.6
kombu==5.0.2
prompt-toolkit==3.0.15
pytz==2021.1
six==1.15.0
vine==5.0.0
wcwidth==0.2.5
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
import celery
from celery import shared_task

@shared_task
def do_something():
    return


@shared_task
def prep_work_before_retry():
    return


@shared_task(bind=True)
def retry_once(self):
    if self.request.retries == 0:
        retry_this_task = self.signature_from_request(
            self.request,
            retries=1,
            immutable=True,
        )

        celery.chain(
            prep_work_before_retry.si(),
            retry_this_task,
        ).delay()

        raise Retry()


# then call:
celery.chain(retry_once.si(), do_something.si()).delay()
```

</p>
</details>

# Expected Behavior

It's expected that the originally delayed chain is executed fully after the retried task succeeded:
- Runs `retry_once` --> raises Retry
- Runs `prep_work_before_retry`
- Re-runs `retry_once`
- Runs  `do_something`

# Actual Behavior

The tasks in the originally delayed chain after the retried task are lost:
- Runs `retry_once` --> raises Retry
- Runs `prep_work_before_retry`
- Re-runs `retry_once`
- Does **NOT** run `do_something`

# Potential source of the issue:

I think these parts are involved in where the rest of the chain is lost:
```python
# snippet from from celery.app.trace.build_tracer, line 463ff.
                        # execute first task in chain
                        chain = task_request.chain
                        if chain:
                            _chsig = signature(chain.pop(), app=app)
                            _chsig.apply_async(
                                (retval,), chain=chain,
                                parent_id=uuid, root_id=root_id,
                                priority=task_priority
                            )
                        mark_as_done(
                            uuid, retval, task_request, publish_result,
                        )

# snippet from celery.canvas.Signature.apply_async

options = {k: v for k, v in options.items() if v is not None}
        try:
            _apply = self._apply_async
        except IndexError:  # pragma: no cover
            # no tasks for chain, etc to find type
            return
        # For callbacks: extra args are prepended to the stored args.
        if args or kwargs or options:
            args, kwargs, options = self._merge(args, kwargs, options)
        else:
            args, kwargs, options = self.args, self.kwargs, self.options
```

**Comment by auvipy** [comment]
did you figure out a possible fix?

**Comment by nbrede** [comment]
A possible fix might be to merge the chains in `celery.canvas.Signature.apply_async`, but I did not dig deep enough to verify whether this would be behavior consistent with other parts and whether that would negatively affect sth else.

Sth like this:
```python
# in celery.canvas.Signature.apply_async, before setting `options`

    sig_chain = self.options.get("chain")
    
    if sig_chain and "chain" in options:
        options_chain = options.get("chain")
        if options_chain:
            sig_chain.extend(options_chain)
        else:
            del options["chain"]

    # continue with current code:
    
    options = {k: v for k, v in options.items() if v is not None}
        try:
            _apply = self._apply_async
        except IndexError:  # pragma: no cover
            # no tasks for chain, etc to find type
            return
        # For callbacks: extra args are prepended to the stored args.
        if args or kwargs or options:
            args, kwargs, options = self._merge(args, kwargs, options)
        else:
            args, kwargs, options = self.args, self.kwargs, self.options

```

**Comment by auvipy** [comment]
@maybe-sybr when you have some time feel free to share insight

**Comment by maybe-sybr** [comment]
Hey, sorry for the delay in responding to this. I've had a read through and I'm not sure the expected behaviour should quite be as @nbrede specified in the original description. The way I'm reading the logic, it seems like the expected behaviour should be:

 * 1: `retry_once()` runs for the first time and delays the new chain of `prep_work | retry_once(retries=1)` then raises `Retry`
 * Now we have two tasks in flight which can be run, followed as A/B below
 * 2A. `retry_once()` is fired again because of the raised `Retry` with `retries=1` and does nothing, returning `None`
 * 2B. `pre_work()` is fired, does nothing and returns `None`
 * 3A. `do_something()` **should** fire with the result of the retried original signature of `retry_once()` (not the reconstructed signature which was delayed by `retry_once()` the first time around)
 * 3B. `retry_once()` is fired from the delayed reconstructed signature with the result from `pre_work()` (`None`), does nothing and returns `None`
 * 4. Both chains are done

Does that sound right to you, @nbrede , or am I missing some subtlety? IMO if you wanted the result of `pre_work | retry_once` to get to `do_something()`, you'd need to be doing a `self.replace()` at the very least.

In any case, it's surprising to me that you aren't seeing `do_something()` firing at least once, even if it's not from the execution of `retry_once()` which you appear to expect. I'm not in a position to run or mess with your test case right now, but should be able to find some time later in the week to have a look.

**Comment by nbrede** [comment]
Hi @maybe-sybr, thanks for looking into this!

> Edit: Oh, perhaps what I missed is that you're creating a new signature from self.request, which should have the remainder of the chain in it. I think I understand...

correct

> Does that sound right to you, @nbrede , or am I missing some subtlety? IMO if you wanted the result of pre_work | retry_once to get to do_something(), you'd need to be doing a self.replace() at the very least.

I wasn't aware of `Task.replace`, not sure yet whether what I want to achieve would work (more easily) with that.

I guess you already have a different understanding based on your edit. I still think what I outlined as expected behavior is what I would expect.

Specifically:

> 2A. retry_once() is fired again because of the raised Retry with retries=1 and does nothing, returning None

I might be wrong, but I think this **SHOULD NOT** happen, as I am raising the `Retry` exception and **NOT** `self.retry` / `Task.retry`.

Instead, I only expect:

> 2B. pre_work() is fired, does nothing and returns None

and then

> 3B. retry_once() is fired from the delayed reconstructed signature with the result from pre_work() (None), does nothing and returns None

followed by `do_something()`, from the reconstruncted signature.

**Comment by thedrow** [comment]
The `self.retry()` method raises a `Retry` exception unless you set the `throw` keyword argument to `False`.
See [here](https://github.com/celery/celery/blob/master/celery/app/task.py#L715-L730).

**Comment by nbrede** [comment]
@thedrow yes, thanks, but I am not calling the `self.retry()` method, but raising the `Retry` exception in the sample code.

**Comment by thedrow** [comment]
I don't understand the difference.

**Comment by nbrede** [comment]
> I don't understand the difference.

Calling `self.retry()` will call `apply_async` on a new instance of the signature again (among other things depending on the exact situation) and then may raise a `Retry` exception.

Raising a `Retry` exception will only ensure the right event is sent/status of the task is set, while it will NOT `apply_async` the signature. As you can see in the MRTC I do that manually to achieve the desired execution of `prep_work_before_retry` before retrying the task again.

Is the intention of the MRTC clear now?

**Comment by maybe-sybr** [comment]
After thinking a bit about this overnight, I was a bit concerned about something trying to be helpful in `app/trace.py` and replacing your carefully constructed retry signature, but it doesn't seem like that's the case at all.

It looks like everything is functioning as you explain, thanks for the clarifications. The point about the difference between raising a `Retry` yourself; which is effectively a signal to the trace logic that a retry should be happening, without using the actual re-scheduling logic from `self.retry()` FWICT; is very helpful.

Here is a slightly modified MRTC which I used to confirm that the correct retries value was being used for the second execution of `retry_once()`:

```python
import celery
from celery import shared_task
from celery.exceptions import Retry

app = celery.Celery(
    broker='amqp://127.13.37.1',
    backend='redis://127.13.37.1',
)

@shared_task
def do_something(*a):
    return 42


@shared_task
def prep_work_before_retry(*a):
    return 1337


@shared_task(bind=True)
def retry_once(self, *a):
    if self.request.retries == 0:
        retry_this_task = self.signature_from_request(
            self.request,
            retries=42,
            immutable=True,
        )
        print(f"{retry_this_task=}")
        print(f"{dict(retry_this_task)=}")

        c = celery.chain(
            prep_work_before_retry.si(),
            retry_this_task,
        )
        print(f"{c=}")
        print(f"{dict(c)=}")
        print(f"{c.tasks[-1]=}")
        print(f"{dict(c.tasks[-1])=}")
        c.delay()

        raise Retry()
    return 31337


# then call:
if __name__ == '__main__':
    celery.chain(retry_once.si(), do_something.si()).delay()
    app.start()
```

-> 

```
[2021-04-13 10:27:51,905: INFO/MainProcess] celery@karma ready.
[2021-04-13 10:27:51,905: INFO/MainProcess] Received task: __main__.retry_once[c2877d0f-f889-4fe2-9997-cad2d5558079]
[2021-04-13 10:27:52,008: WARNING/ForkPoolWorker-7] retry_this_task=__main__.retry_once()
[2021-04-13 10:27:52,008: WARNING/ForkPoolWorker-7] dict(retry_this_task)={'task': '__main__.retry_once', 'args': (), 'kwargs': {}, 'options': {'task_id': 'c2877d0f-f889-4fe2-9997-cad2d5558079', 'root_id': 'c2877d0f-f889-4fe2-9997-cad2d5558079', 'parent_id': None, 'group_id': None, 'group_index': None, 'shadow': None, 'chord': None, 'chain': [{'task': '__main__.do_something', 'args': [], 'kwargs': {}, 'options': {'task_id': '74042373-0182-416f-a86e-b83d4f331304', 'reply_to': '966f1f00-96a0-35c0-89a8-9fdde10def21'}, 'subtask_type': None, 'immutable': True}], 'link': None, 'link_error': None, 'expires': None, 'soft_time_limit': None, 'time_limit': None, 'headers': None, 'retries': 42, 'reply_to': '966f1f00-96a0-35c0-89a8-9fdde10def21', 'origin': 'gen2528662@karma', 'priority': 0, 'queue': 'celery'}, 'subtask_type': None, 'immutable': True}
[2021-04-13 10:27:52,009: WARNING/ForkPoolWorker-7] c=__main__.prep_work_before_retry() | retry_once()
[2021-04-13 10:27:52,009: WARNING/ForkPoolWorker-7] dict(c)={'task': 'celery.chain', 'args': (), 'kwargs': {'tasks': [__main__.prep_work_before_retry(), __main__.retry_once()]}, 'options': {}, 'subtask_type': 'chain', 'immutable': False}
[2021-04-13 10:27:52,010: WARNING/ForkPoolWorker-7] c.tasks[-1]=__main__.retry_once()
[2021-04-13 10:27:52,010: WARNING/ForkPoolWorker-7] dict(c.tasks[-1])={'task': '__main__.retry_once', 'args': (), 'kwargs': {}, 'options': {'task_id': 'c2877d0f-f889-4fe2-9997-cad2d5558079', 'root_id': 'c2877d0f-f889-4fe2-9997-cad2d5558079', 'parent_id': None, 'group_id': None, 'group_index': None, 'shadow': None, 'chord': None, 'chain': [{'task': '__main__.do_something', 'args': [], 'kwargs': {}, 'options': {'task_id': '74042373-0182-416f-a86e-b83d4f331304', 'reply_to': '966f1f00-96a0-35c0-89a8-9fdde10def21'}, 'subtask_type': None, 'immutable': True}], 'link': None, 'link_error': None, 'expires': None, 'soft_time_limit': None, 'time_limit': None, 'headers': None, 'retries': 42, 'reply_to': '966f1f00-96a0-35c0-89a8-9fdde10def21', 'origin': 'gen2528662@karma', 'priority': 0, 'queue': 'celery'}, 'subtask_type': None, 'immutable': True}
[2021-04-13 10:27:52,020: INFO/MainProcess] Received task: __main__.prep_work_before_retry[49cb6ce8-3445-45eb-acdc-1579be35926d]
[2021-04-13 10:27:52,021: INFO/ForkPoolWorker-7] Task __main__.retry_once[c2877d0f-f889-4fe2-9997-cad2d5558079] retry: Retry at None
[2021-04-13 10:27:52,029: INFO/MainProcess] Received task: __main__.retry_once[c2877d0f-f889-4fe2-9997-cad2d5558079]
[2021-04-13 10:27:52,030: INFO/ForkPoolWorker-8] Task __main__.prep_work_before_retry[49cb6ce8-3445-45eb-acdc-1579be35926d] succeeded in 0.008860207000907394s: 1337
[2021-04-13 10:27:52,030: INFO/ForkPoolWorker-7] Task __main__.retry_once[c2877d0f-f889-4fe2-9997-cad2d5558079] succeeded in 0.0005692630002158694s: 31337
^C
worker: Hitting Ctrl+C again will terminate all running tasks!
```

So yeah, the signature object you're delaying looks entirely sensible to me. This must be some misbehaviour in `_apply_async()` as you've identified in the top post.

To hopefully unblock your actual problem which led to the report, it seems like doing `raise self.replace(the_new_chain)` instead of `the_new_chain.delay() ; raise Retry()` runs the tasks as you would expect. Diff to my MRTC above and debug output so you can sanity check the args get passed around as expected (the sigs are not immutable in this run):

```diff
1d0
< 
33c32
<             prep_work_before_retry.si(),
---
>             prep_work_before_retry.s(),
40,42c39
<         c.delay()
< 
<         raise Retry()
---
>         raise self.replace(c)
48c45
<     celery.chain(retry_once.si(), do_something.si()).delay()
---
>     celery.chain(retry_once.s(), do_something.s()).delay()
```
->
```
[2021-04-13 10:35:45,346: INFO/MainProcess] celery@karma ready.
[2021-04-13 10:35:45,346: DEBUG/MainProcess] basic.qos: prefetch_count->32
[2021-04-13 10:35:45,346: INFO/MainProcess] Received task: __main__.retry_once[de03545d-3996-4a8a-a889-c40ade170f12]
[2021-04-13 10:35:45,346: DEBUG/MainProcess] TaskPool: Apply <function fast_trace_task at 0x7f86880f0790> (args:('__main__.retry_once', 'de03545d-3996-4a8a-a889-c40ade170f12', {'lang': 'py', 'task': '__main__.retry_once', 'id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'shadow': None, 'eta': None, 'expires': None, 'group': None, 'group_index': None, 'retries': 0, 'timelimit': [None, None], 'root_id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'parent_id': None, 'argsrepr': '()', 'kwargsrepr': '{}', 'origin': 'gen2530882@karma', 'reply_to': 'e480f93e-ac0c-3ed2-a931-bfb144f6fcc6', 'correlation_id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'hostname': 'celery@karma', 'delivery_info': {'exchange': '', 'routing_key': 'celery', 'priority': 0, 'redelivered': False}, 'args': [], 'kwargs': {}}, '[[], {}, {"callbacks": null, "errbacks": null, "chain": [{"task": "__main__.do_something", "args": [], "kwargs": {}, "options": {"task_id": "d229d5fd-c195-4fb2-bc25-067232f49a75", "reply_to": "e480f93e-ac0c-3ed2-a931-bfb144f6fcc6"}, "subtask_type": null, "immutable": false}], "chord": null}]', 'application/json', 'utf-8') kwargs:{})
[2021-04-13 10:35:45,449: WARNING/ForkPoolWorker-7] retry_this_task=__main__.retry_once()
[2021-04-13 10:35:45,449: WARNING/ForkPoolWorker-7] dict(retry_this_task)={'task': '__main__.retry_once', 'args': (), 'kwargs': {}, 'options': {'task_id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'root_id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'parent_id': None, 'group_id': None, 'group_index': None, 'shadow': None, 'chord': None, 'chain': [{'task': '__main__.do_something', 'args': [], 'kwargs': {}, 'options': {'task_id': 'd229d5fd-c195-4fb2-bc25-067232f49a75', 'reply_to': 'e480f93e-ac0c-3ed2-a931-bfb144f6fcc6'}, 'subtask_type': None, 'immutable': False}], 'link': None, 'link_error': None, 'expires': None, 'soft_time_limit': None, 'time_limit': None, 'headers': None, 'retries': 42, 'reply_to': 'e480f93e-ac0c-3ed2-a931-bfb144f6fcc6', 'origin': 'gen2530882@karma', 'priority': 0, 'queue': 'celery'}, 'subtask_type': None, 'immutable': True}
[2021-04-13 10:35:45,450: WARNING/ForkPoolWorker-7] c=__main__.prep_work_before_retry() | retry_once()
[2021-04-13 10:35:45,450: WARNING/ForkPoolWorker-7] dict(c)={'task': 'celery.chain', 'args': (), 'kwargs': {'tasks': [__main__.prep_work_before_retry(), __main__.retry_once()]}, 'options': {}, 'subtask_type': 'chain', 'immutable': False}
[2021-04-13 10:35:45,450: WARNING/ForkPoolWorker-7] c.tasks[-1]=__main__.retry_once()
[2021-04-13 10:35:45,451: WARNING/ForkPoolWorker-7] dict(c.tasks[-1])={'task': '__main__.retry_once', 'args': (), 'kwargs': {}, 'options': {'task_id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'root_id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'parent_id': None, 'group_id': None, 'group_index': None, 'shadow': None, 'chord': None, 'chain': [{'task': '__main__.do_something', 'args': [], 'kwargs': {}, 'options': {'task_id': 'd229d5fd-c195-4fb2-bc25-067232f49a75', 'reply_to': 'e480f93e-ac0c-3ed2-a931-bfb144f6fcc6'}, 'subtask_type': None, 'immutable': False}], 'link': None, 'link_error': None, 'expires': None, 'soft_time_limit': None, 'time_limit': None, 'headers': None, 'retries': 42, 'reply_to': 'e480f93e-ac0c-3ed2-a931-bfb144f6fcc6', 'origin': 'gen2530882@karma', 'priority': 0, 'queue': 'celery'}, 'subtask_type': None, 'immutable': True}
[2021-04-13 10:35:45,461: DEBUG/ForkPoolWorker-7] Start from server, version: 0.9, properties: {'capabilities': {'publisher_confirms': True, 'exchange_exchange_bindings': True, 'basic.nack': True, 'consumer_cancel_notify': True, 'connection.blocked': True, 'consumer_priorities': True, 'authentication_failure_close': True, 'per_consumer_qos': True, 'direct_reply_to': True}, 'cluster_name': 'rabbit@01fdea29c0cd', 'copyright': 'Copyright (c) 2007-2021 VMware, Inc. or its affiliates.', 'information': 'Licensed under the MPL 2.0. Website: https://rabbitmq.com', 'platform': 'Erlang/OTP 23.3.1', 'product': 'RabbitMQ', 'version': '3.8.14'}, mechanisms: [b'PLAIN', b'AMQPLAIN'], locales: ['en_US']
[2021-04-13 10:35:45,462: DEBUG/ForkPoolWorker-7] using channel_id: 1
[2021-04-13 10:35:45,463: DEBUG/ForkPoolWorker-7] Channel open
[2021-04-13 10:35:45,465: INFO/MainProcess] Received task: __main__.prep_work_before_retry[e3642952-16d7-4dc3-8941-c56a565a8b0d]
[2021-04-13 10:35:45,465: DEBUG/MainProcess] TaskPool: Apply <function fast_trace_task at 0x7f86880f0790> (args:('__main__.prep_work_before_retry', 'e3642952-16d7-4dc3-8941-c56a565a8b0d', {'lang': 'py', 'task': '__main__.prep_work_before_retry', 'id': 'e3642952-16d7-4dc3-8941-c56a565a8b0d', 'shadow': None, 'eta': None, 'expires': None, 'group': None, 'group_index': None, 'retries': 0, 'timelimit': [None, None], 'root_id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'parent_id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'argsrepr': '()', 'kwargsrepr': '{}', 'origin': 'gen2530890@karma', 'reply_to': 'e480f93e-ac0c-3ed2-a931-bfb144f6fcc6', 'correlation_id': 'e3642952-16d7-4dc3-8941-c56a565a8b0d', 'hostname': 'celery@karma', 'delivery_info': {'exchange': '', 'routing_key': 'celery', 'priority': 0, 'redelivered': False}, 'args': [], 'kwargs': {}}, '[[], {}, {"callbacks": null, "errbacks": null, "chain": [{"task": "__main__.do_something", "args": [], "kwargs": {}, "options": {"task_id": "d229d5fd-c195-4fb2-bc25-067232f49a75", "reply_to": "e480f93e-ac0c-3ed2-a931-bfb144f6fcc6", "root_id": "de03545d-3996-4a8a-a889-c40ade170f12"},... kwargs:{})
[2021-04-13 10:35:45,465: INFO/ForkPoolWorker-7] Task __main__.retry_once[de03545d-3996-4a8a-a889-c40ade170f12] ignored
[2021-04-13 10:35:45,473: DEBUG/ForkPoolWorker-8] Start from server, version: 0.9, properties: {'capabilities': {'publisher_confirms': True, 'exchange_exchange_bindings': True, 'basic.nack': True, 'consumer_cancel_notify': True, 'connection.blocked': True, 'consumer_priorities': True, 'authentication_failure_close': True, 'per_consumer_qos': True, 'direct_reply_to': True}, 'cluster_name': 'rabbit@01fdea29c0cd', 'copyright': 'Copyright (c) 2007-2021 VMware, Inc. or its affiliates.', 'information': 'Licensed under the MPL 2.0. Website: https://rabbitmq.com', 'platform': 'Erlang/OTP 23.3.1', 'product': 'RabbitMQ', 'version': '3.8.14'}, mechanisms: [b'PLAIN', b'AMQPLAIN'], locales: ['en_US']
[2021-04-13 10:35:45,473: DEBUG/ForkPoolWorker-8] using channel_id: 1
[2021-04-13 10:35:45,474: DEBUG/ForkPoolWorker-8] Channel open
[2021-04-13 10:35:45,476: INFO/MainProcess] Received task: __main__.retry_once[de03545d-3996-4a8a-a889-c40ade170f12]
[2021-04-13 10:35:45,476: DEBUG/MainProcess] TaskPool: Apply <function fast_trace_task at 0x7f86880f0790> (args:('__main__.retry_once', 'de03545d-3996-4a8a-a889-c40ade170f12', {'lang': 'py', 'task': '__main__.retry_once', 'id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'shadow': None, 'eta': None, 'expires': None, 'group': None, 'group_index': None, 'retries': 42, 'timelimit': [None, None], 'root_id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'parent_id': 'e3642952-16d7-4dc3-8941-c56a565a8b0d', 'argsrepr': '()', 'kwargsrepr': '{}', 'origin': 'gen2530891@karma', 'reply_to': 'e480f93e-ac0c-3ed2-a931-bfb144f6fcc6', 'correlation_id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'hostname': 'celery@karma', 'delivery_info': {'exchange': '', 'routing_key': 'celery', 'priority': 0, 'redelivered': False}, 'args': [], 'kwargs': {}}, '[[], {}, {"callbacks": null, "errbacks": null, "chain": [{"task": "__main__.do_something", "args": [], "kwargs": {}, "options": {"task_id": "d229d5fd-c195-4fb2-bc25-067232f49a75", "reply_to": "e480f93e-ac0c-3ed2-a931-bfb144f6fcc6", "root_id": "de03545d-3996-4a8a-a889-c40ade170f12"}, "subtask_type": null,... kwargs:{})
[2021-04-13 10:35:45,476: INFO/ForkPoolWorker-8] Task __main__.prep_work_before_retry[e3642952-16d7-4dc3-8941-c56a565a8b0d] succeeded in 0.0109349119993567s: 1337
[2021-04-13 10:35:45,477: INFO/MainProcess] Received task: __main__.do_something[d229d5fd-c195-4fb2-bc25-067232f49a75]
[2021-04-13 10:35:45,478: DEBUG/MainProcess] TaskPool: Apply <function fast_trace_task at 0x7f86880f0790> (args:('__main__.do_something', 'd229d5fd-c195-4fb2-bc25-067232f49a75', {'lang': 'py', 'task': '__main__.do_something', 'id': 'd229d5fd-c195-4fb2-bc25-067232f49a75', 'shadow': None, 'eta': None, 'expires': None, 'group': None, 'group_index': None, 'retries': 0, 'timelimit': [None, None], 'root_id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'parent_id': 'de03545d-3996-4a8a-a889-c40ade170f12', 'argsrepr': '(31337,)', 'kwargsrepr': '{}', 'origin': 'gen2530890@karma', 'reply_to': 'e480f93e-ac0c-3ed2-a931-bfb144f6fcc6', 'correlation_id': 'd229d5fd-c195-4fb2-bc25-067232f49a75', 'hostname': 'celery@karma', 'delivery_info': {'exchange': '', 'routing_key': 'celery', 'priority': 0, 'redelivered': False}, 'args': [31337], 'kwargs': {}}, '[[31337], {}, {"callbacks": null, "errbacks": null, "chain": [], "chord": null}]', 'application/json', 'utf-8') kwargs:{})
[2021-04-13 10:35:45,478: INFO/ForkPoolWorker-7] Task __main__.retry_once[de03545d-3996-4a8a-a889-c40ade170f12] succeeded in 0.0015327460005210014s: 31337
[2021-04-13 10:35:45,478: INFO/ForkPoolWorker-7] Task __main__.do_something[d229d5fd-c195-4fb2-bc25-067232f49a75] succeeded in 0.0003747869996004738s: 42
^C
worker: Hitting Ctrl+C again will terminate all running tasks!
```

Hope this helps in the meantime until we have a patch for the breakage :)

**Comment by nbrede** [comment]
Thanks a lot for looking into this and your suggested workaround, will try that!
